pkg/registry/core/service/strategy.go

```diff
@@ -205,6 +210,17 @@ func topologyKeysInUse(svc *api.Service) bool {
 	return len(svc.Spec.TopologyKeys) > 0
 }
 
+// returns true if svc.Spec.ServiceIPFamily field is in use
+func allocateLoadBalancerNodePortsInUse(svc *api.Service) bool {
+	if svc == nil {
+		return false
+	}
+	if svc.Spec.AllocateLoadBalancerNodePorts != nil && !*svc.Spec.AllocateLoadBalancerNodePorts {
```

Comment:
I _think_ the "in use" check should only check if the field is not nil and not the actual bool value.

Reply:
The reasoning was that if a user sets AllocateLoadBalancerNodePorts=true (a no-op) when the feature-gate is enabled then the AllocateLoadBalancerNodePorts would still be visible when doing "kubectl get service -o json ..." even when the feature-gate is disabled. With the current condition it would become hidden again.

Reply:
But when the gate is disabled we would nil out the field in line 190 so it would be hidden again right?

Reply:
No, it would not be set to nil since the "inUse" function will return true unless it is already nil. Compare to the `serviceIPFamilyInUse` function which will make sure the value is not set to `nil` unless it already is `nil` which makes only the svc == nil check useful.

Reply:
Oh I see -- so if the gate is disabled but AllocateLoadBalanceNodePort was set to a value that is not the default (false) we shouldn't nil it. 

But even if it was explicitly set to `true` shouldn't we leave it instead of setting it to nil? The current logic would return `false` here if `AllocateLoadBalanceNodePort = true` which would result in a nil field on line 190.

Reply:
In the case of dual-stack, we only check if IPFamily is nil, we don't care what the value is.

https://github.com/kubernetes/kubernetes/blob/db28b0239a1d09e46507e0b320f0b4ffa1cbc8b6/pkg/registry/core/service/strategy.go#L194-L196

Reply:
> so if the gate is disabled but AllocateLoadBalanceNodePort was set to a value that is not the default (false) we shouldn't nil it.

Yes. Compare with the `topologyKeysInUse` function. The value is not set to nil if there are defined values.

The `serviceIPFamilyInUse()` on the other hand is useless. It returns true only if the value is != nil and the negated condition `!serviceIPFamilyInUse(oldSvc)` at line 179 make sure that the value is only set to nil it is _already nil_. I think the intended function would be to return true (indicating "in-use") only if the ipFamily was different from the "main" family (default). But this will change soon anyway when the ipFamilies becomes an array.

Reply:
> Yes. Compare with the topologyKeysInUse function. The value is not set to nil if there are defined values.

Right, so I think we're on the same page in that we shouldn't set the field to nil if it is already set. But I think when the field is set to `true`, we should leave the field as `true` as opposed to setting it to nil. In order to do this, we need to change this if condition to:

```go
if svc.Spec.AllocateLoadBalancerNodePorts != nil {
    return true
}
```

Otherwise on rollback we would drop the field that was set. Although behaviourally there is no difference, it would still be incorrect to drop that field. If `nil` and `true` that different behaviors (not in this case, but true for other fields), we would only do a `nil` check here as well.

Reply:
With that condition the function is useless since it _will_ only set the field to nil if it already is nil. Can I remove the function entirely?

Reply:
Don't think it hurts to have this function just to be consistent with other fields. Maybe it can be simplified to:
```
func allocateLoadBalancerNodePortsInUse(svc *api.Service) bool {
	if svc == nil {
		return false
	}

        return svc.Spec.AllocateLoadBalancerNodePorts != nil
}
```

Reply:
I'll update with this as you suggest soon. There may be some case I have missed and it is in line with other similar functions.

I will just rebase and re-generate...